src/bytes_mut.rs

```diff
@@ -819,6 +820,73 @@ impl BytesMut {
         }
     }
 
+    /// Attempts to reclaim the whole allocation of the `BytesMut`.
+    ///
+    /// If the `BytesMut` is empty but the underlying storage has been used before,
+    /// it might be possible to cheaply reclaim space by just updating a few indices.
+    /// Returns `None` if the `BytesMut` is not empty, there is nothing to reclaim
+    /// (no underlying storage has been allocated), or there are any other live
+    /// references to the underlying storage. Otherwise, returns the available
+    /// capacity after reclaiming.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bytes::BytesMut;
+    /// use core::num::NonZeroUsize;
+    ///
+    /// let mut buf = BytesMut::with_capacity(64);
+    /// assert_eq!(None, buf.try_reclaim());
+    ///
+    /// buf.extend_from_slice(b"abcd");
+    /// let mut split = buf.split();
+    /// assert_eq!(None, split.try_reclaim());
+    /// assert_eq!(None, buf.try_reclaim());
+    /// drop(buf);
+    /// assert_eq!(None, split.try_reclaim());
+    /// split.clear();
+    /// assert_eq!(Some(64), split.try_reclaim().map(NonZeroUsize::into));
+    /// ```
+    pub fn try_reclaim(&mut self) -> Option<NonZeroUsize> {
```

Comment:
I'm fine adding a reclaim API (if there is no way to perform the OP today). I don't love the `Option<NonZeroUsize>` return value. It seems to me it could just return the capacity after the fact regardless of whether or not any more space was reclaimed?

Reply:
I don't know a way to do it without potentially allocating. I am not much attached to the return value, I just would like some way to check whether reclaim actually did anything. Comparing the capacity before and after the call is not really viable for that case, as the underlying buffer might legitimately be zero bytes long. Would returning a bool to indicate whether anything was reclaimed be more to your taste?

Reply:
Returning a `bool` makes sense to me. We can mark it `#[must_use]` as well.